modules/core/src/main/java/org/apache/ignite/binary/BinaryBasicNameMapper.java

```diff
@@ -87,6 +87,18 @@ public BinaryBasicNameMapper setSimpleName(boolean isSimpleName) {
     private static String simpleName(String clsName) {
         assert clsName != null;
 
+        // .NET generic, not valid for Java class name. Clean up every generic part recursively.
+        // Example: Foo.Bar`1[[Baz.Qux`1[[System.String],[System.Int32]]]]
+        int genericIdx = clsName.indexOf("[[");
+
+        if (genericIdx > 0)
+            clsName = clsName.substring(0, genericIdx + 2) + simpleName(clsName.substring(genericIdx + 2));
+
+        genericIdx = clsName.indexOf("],[", genericIdx);
+
+        if (genericIdx > 0)
+            clsName = clsName.substring(0, genericIdx + 3) + simpleName(clsName.substring(genericIdx + 3));
+
```

Comment:
Consider extract this as a separate function that can be tested separately on simple strings.

Reply:
Good point, I've extracted a method. However, I don't think we should test it separately - it is an implementation detail. Corresponding tests added to `BinaryBasicNameMapperSelfTest`.